### PR TITLE
[Sprint 62] XD-3216-3610 Offset management and header propagation issues

### PR DESCRIPTION
**Pull request by mbogoevici**
* Upgrade to Spring Integration Kafka 1.3.0.RELEASE and Kafka 0.8.2.2
* Add an option for native offset management for Kafka sources and make it the default
* Add native offset management support to Kafka message bus, make it configurable and default

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -116,6 +116,7 @@ allprojects {
 		mavenCentral()
 		maven { url 'http://repo.spring.io/milestone' }
 		maven { url 'http://repo.spring.io/libs-milestone' }
+		maven { url 'http://repo.spring.io/libs-staging-local' }
```
Comment (mbogoevici) [inline build.gradle]: To be removed once SI Kafka 1.3.0 is fully released.

Comment (garyrussell) [inline build.gradle]: @mbogoevici - I have promoted the 1.3.0.RELEASE

**Comment by mbogoevici** [comment]
Pushed update removing the staging repo.

File: extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java [hunk 2]
```diff
@@ -175,7 +175,8 @@ public int getQueueSize() {
 	public enum OffsetStorageStrategy {
 		inmemory,
 		redis,
-		kafka
+		kafka,
+		kafka_native
```
Comment (garyrussell) [inline extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java]: Perhaps it should be camelCase to be consistent with the bus property?

File: spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java [hunk 3]
```diff
@@ -812,21 +815,30 @@ private KafkaMessageListenerContainer createMessageListenerContainer(KafkaProper
...
-			kafkaOffsetManager.setRequiredAcks(offsetStoreRequiredAcks);
-			kafkaOffsetManager.setMaxSize(offsetStoreMaxFetchSize);
-			kafkaOffsetManager.setBatchBytes(offsetStoreBatchBytes);
-			kafkaOffsetManager.setMaxQueueBufferingTime(offsetStoreBatchTime);
-
-			kafkaOffsetManager.afterPropertiesSet();
-
-			WindowingOffsetManager windowingOffsetManager = new WindowingOffsetManager(kafkaOffsetManager);
+			OffsetManager delegateOffsetManager;
+			if (OffsetManagement.kafkaNative.equals(offsetManagement)) {
```
Comment (garyrussell) [inline spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java]: Logic reversed.

File: extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java (hunk 2 above)
Comment (mbogoevici) [inline extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaSourceModuleOptionsMetadata.java]: Makes sense.

File: spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java (hunk 3 above)
Comment (mbogoevici) [inline spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java]: Right.

Comment (garyrussell) [inline spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java]: Also, no setter - needs a unit test.

Comment (mbogoevici) [inline spring-xd-messagebus-kafka/src/main/java/org/springframework/xd/dirt/integration/kafka/KafkaMessageBus.java]: Ok, I'll add one.

File: src/docs/asciidoc/Sources.asciidoc [hunk 4]
```diff
@@ -458,7 +458,7 @@ $$kafkaOffsetTopicName$$:: $$name of the offset topic, if Kafka offset strategy
 $$kafkaOffsetTopicRequiredAcks$$:: $$required acks for writing to the Kafka offset topic, if Kafka offset strategy is chosen$$ *($$int$$, default: `1`)*
 $$kafkaOffsetTopicRetentionTime$$:: $$retention time for dead records (tombstones), if Kafka offset strategy is chosen$$ *($$int$$, default: `60000`)*
 $$kafkaOffsetTopicSegmentSize$$:: $$segment size of the offset topic, if Kafka offset strategy is chosen$$ *($$int$$, default: `262144000`)*
-$$offsetStorage$$:: $$strategy for persisting offset values$$ *($$OffsetStorageStrategy$$, default: `kafka`, possible values: `inmemory,redis,kafka`)*
+$$offsetStorage$$:: $$strategy for persisting offset values$$ *($$OffsetStorageStrategy$$, default: `kafka_native`, possible values: `inmemory,redis,kafka,kafka_native`)*
```
Comment (artembilan) [inline src/docs/asciidoc/Sources.asciidoc]: So, this one to camelCase as well.

Comment (ericbottard) [inline src/docs/asciidoc/Sources.asciidoc]: This is generated and will change automatically

Comment (artembilan) [inline src/docs/asciidoc/Sources.asciidoc]: Good. Thanks.
So, what is the point of this noise here for review then?
Why can't it be generated during the `reference` task ?

Comment (ericbottard) [inline src/docs/asciidoc/Sources.asciidoc]: Because it used to be in the wiki. Doesn't really make sense now, I agree.

**Comment by mbogoevici** [comment]
Pushed a rebased version + polishing.

**Comment by garyrussell** [comment]
Merged as 993b0b1ddc9672325044c8a440bdfef7147692c8

**Pull request closed without merging**